src/dachshund/cnm_communities.rs

```diff
@@ -0,0 +1,303 @@
+                }
+            };
+            new_delta_qjk_map.insert(k, new_delta_qjk);
+
+            /* Update the binary maps for k */
+            let neighbors_k: &mut HashMap<usize, f64> = delta_q_bmap.get_mut(&k).unwrap();
+            if delta_qik.is_some() {
+                neighbors_k.remove(&i);
+            }
+            neighbors_k.insert(j, new_delta_qjk);
+
+            /* Update the binary heap for k */
+            let old_maxheap: CNMCommunityMergeInstructionHeap = delta_q_maxheap.remove(&k).unwrap();
+            let mut new_maxheap: CNMCommunityMergeInstructionHeap =
+                BinaryHeap::with_capacity(old_maxheap.len());
+            for el in old_maxheap.into_iter_sorted() {
+                let ll = el.j;
+
+                if ll != i {
+                    if ll == j {
+                        new_maxheap.push(CNMCommunityMergeInstruction::new(
+                            OrderedFloat(new_delta_qjk),
+                            k,
+                            ll,
+                        ));
+                    } else {
+                        new_maxheap.push(el);
+                    }
+                }
+            }
+            delta_q_maxheap.insert(k, new_maxheap);
+            new_community_maxheap.push(CNMCommunityMergeInstruction::new(
+                OrderedFloat(new_delta_qjk),
+                j,
+                k,
+            ));
+        }
+        // adding the new_delta_qjk map for the newly created community
+        delta_q_bmap.insert(j, new_delta_qjk_map);
+        delta_q_bmap.remove(&i);
+        // updating the delta_q_maxheap
+        delta_q_maxheap.insert(j, new_community_maxheap);
+        delta_q_maxheap.remove(&i);
+
+        // updating the degree map
+        let new_degree = degree_map[&i] + degree_map[&j];
+        degree_map.insert(j, new_degree);
+        degree_map.remove(&i);
+
+        maxh = self.get_max_maxheap(&delta_q_maxheap);
+        (
+            communities,
+            degree_map,
+            delta_q_bmap,
+            delta_q_maxheap,
+            maxh,
+            num_edges,
+        )
+    }
+    fn get_cnm_communities(&self) -> (HashMap<usize, Community>, Vec<f64>) {
+        let (
+            mut communities,
+            mut degree_map,
+            mut delta_q_bmap,
+            mut delta_q_maxheap,
+            mut maxh,
+            num_edges,
+        ) = self.init_cnm_communities();
+
+        let mut modularity_change = maxh.peek().unwrap().delta_ij.into_inner();
+        let mut modularity_changes: Vec<f64> = vec![modularity_change];
+
+        while maxh.len() > 0 && modularity_change > 0. {
+            let res = self.iterate_cnm_communities(
+                communities,
+                degree_map,
+                delta_q_bmap,
+                delta_q_maxheap,
+                maxh,
+                num_edges,
+            );
+            communities = res.0;
```

Comment:
Can you do this with something like let (communities, degree_map, ...) = self.iterate_cnm_communities(...) ? That reads a bit cleaner to me since the res.1 res.2 etc are all just magic numbers anyway.

Reply:
yeah, will actually do the right thing and implement that intermediary state struct :)